fontscan/fontconfig.go

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Unlicense OR BSD-3-Clause
+
+package fontscan
+
+import (
+	"encoding/xml"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Support for a (very limited) subset of the Linux fontconfig config file format
+// See https://www.freedesktop.org/wiki/Software/fontconfig/ for reference
+
+const fcRootConfig = "/etc/fonts/fonts.conf"
+
+const (
+	_ = iota
+	fcDir
+	fcInclude
+)
+
+// fcDirective is either a <dir> or a <include> element,
+// as indicated by [kind]
+type fcDirective struct {
+	dir struct {
+		Dir    string `xml:",chardata"`
+		Prefix string `xml:"prefix,attr"`
+	}
+	include struct {
+		Include       string `xml:",chardata"`
+		IgnoreMissing string `xml:"ignore_missing,attr"`
+		Prefix        string `xml:"prefix,attr"`
+	}
+	kind uint8
+}
+
+func (directive *fcDirective) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	switch start.Name.Local {
+	case "dir":
+		directive.kind = fcDir
+		return d.DecodeElement(&directive.dir, &start)
+	case "include":
+		directive.kind = fcInclude
+		return d.DecodeElement(&directive.include, &start)
+	default:
+		// ignore the element
+		return d.Skip()
+	}
+}
+
+// parseFcFile opens and process a FontConfig config file,
+// returning the font directories to scan and the (optionnal)
+// supplementary config files (or directories) to include.
+func parseFcFile(file, currentWorkingDir string) (fontDirs, includes []string, _ error) {
+	f, err := os.Open(file)
+	if err != nil {
+		return nil, nil, fmt.Errorf("opening fontconfig config file: %s", err)
+	}
+	defer f.Close()
+
+	var config struct {
+		Fontconfig []fcDirective `xml:",any"`
+	}
+	err = xml.NewDecoder(f).Decode(&config)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parsing fontconfig config file: %s", err)
+	}
+
+	// post-process : handle "prefix" attr and use absolute path
+	xdg := os.Getenv("XDG_DATA_HOME")
+	for _, item := range config.Fontconfig {
+		switch item.kind {
+		case fcDir:
+			dir := item.dir.Dir
+			switch item.dir.Prefix {
+			case "default", "cwd":
+				dir = filepath.Join(currentWorkingDir, dir)
+			case "relative":
+				dir = filepath.Join(filepath.Dir(file), dir)
+			case "xdg":
+				dir = filepath.Join(xdg, dir)
+			}
+			fontDirs = append(fontDirs, dir)
+		case fcInclude:
+			include := item.include.Include
+			if item.include.Prefix == "xdg" {
+				include = filepath.Join(xdg, include)
+			} else {
+				// handle implicit relative dirs
+				if strings.HasPrefix(include, "~") {
+					include = expandUser(include)
+				} else if !filepath.IsAbs(include) {
+					include = filepath.Join(filepath.Dir(file), include)
+				}
+			}
+			includes = append(includes, include)
+		}
+	}
+	return
+}
+
+// parseFcDir processes all the files in [dir] matching the [09]*.conf pattern
+// seen is updated with the processed fontconfig files
```

Comment:
I don't think that using a numeric prefix is actually *required* for those files. Users can insert new ones that don't follow the pattern, they just may have unexpected ordering due to lexicographic sorting. We should parse anything with the right extension.

Reply:
According to fontconfig documentation : 

> If a directory, every file within that directory starting with an ASCII digit (U+0030 - U+0039) and ending with the string ``.conf'' will be processed in sorted order. 

Note this is only true for _directories_, not for single files.